### PR TITLE
MM-53409: Remove antijoin from ChannelStore.GetAllChannels

### PR DESCRIPTION
**Pull request by agnivade**
https://mattermost.atlassian.net/browse/MM-53409
```release-note
NONE
```

### CODE REVIEW

**Review by isacikgoz — APPROVED** [review]

File: server/channels/store/sqlstore/channel_store.go [hunk 1]
```diff
@@ -1287,7 +1287,11 @@ func (s SqlChannelStore) getAllChannelsQuery(opts store.ChannelSearchOpts, forCo
 		query = query.LeftJoin("RetentionPoliciesChannels ON c.Id = RetentionPoliciesChannels.ChannelId")
 	}
 	if opts.ExcludePolicyConstrained {
-		query = query.Where("RetentionPoliciesChannels.ChannelId IS NULL")
+		if s.DriverName() == model.DatabaseDriverPostgres {
+			query = query.Where("RetentionPoliciesChannels.ChannelId IS NULL")
+		} else {
+			query = query.Where(sq.Expr("c.Id NOT IN (SELECT ChannelId FROM RetentionPoliciesChannels)"))
+		}
```
Comment (agarciamontoro) [inline server/channels/store/sqlstore/channel_store.go:1294]: Shouldn't we be avoiding the LeftJoin in MySQL?

Comment (agnivade) [reply server/channels/store/sqlstore/channel_store.go:1294]: No, just `LEFT JOIN` is fine. But doing the left join and filtering with `IS NULL` is what causes the antijoin

Comment (agarciamontoro) [reply server/channels/store/sqlstore/channel_store.go:1294]: Oh, I thought the analyses we had all involved avoiding the left join. I don't understand how just changing the `WHERE` condition fixes the plan.

Comment (agnivade) [reply server/channels/store/sqlstore/channel_store.go:1294]: The objective is to remove all antijoins. An antijoin is formed when you do a left join with a table, and filter the rows which are NULL in that table. For example, if you have table A and table B, you do `A LEFT JOIN B ON A.Id=B.UserId WHERE B.UserId IS NULL`, this essentially means you want to get all non-matching rows from table B. That's an antijoin.

However, if you just do `A LEFT JOIN B ON A.Id=B.UserId`, you need all matching + non matching rows. This is a normal join.

Let me know if that helps.

Comment (agarciamontoro) [reply server/channels/store/sqlstore/channel_store.go:1294]: Yeah, that makes sense, but I'm worried we don't have hard data for only removing the `IS NULL` filter. Let me rephrase my question:
My assumption is that the issues we've seen in real-life scenarios occurred only when we had a query of the form 

```sql
SELECT x,y,z FROM A LEFT JOIN B ON A.c=B.d WHERE B.d IS NULL
```

and we changed it by a query of the form

```sql
SELECT x,y,z FROM A WHERE A.c NOT IN (SELECT d from B)
```

If that assumption is correct, I'm a bit concerned that we're in uncharted territory when we only change the `IS NULL` filter by the `NOT IN` one, without removing the `LEFT JOIN`. If that assumption is not correct, then I'm ok with this.

Also, with the current changes, we would have the table `B` (in this case `RetentionPoliciesChannels`) both in the `LEFT JOIN` *and* in the subquery, something like

```sql
SELECT x,y,z FROM A LEFT JOIN B WHERE A.c NOT IN (SELECT d FROM B)
```

Do we even need the left join if we don't use it afterwards and query the joined table again?

Comment (agnivade) [reply server/channels/store/sqlstore/channel_store.go:1294]: You are right. The new query is objectively worse.

OLD:
```
| -> Limit: 10 row(s)  (actual time=1785.084..1785.089 rows=10 loops=1)
    -> Sort: c.DisplayName, Teams.DisplayName, limit input to 10 row(s) per chunk  (actual time=1785.081..1785.086 rows=10 loops=1)
        -> Stream results  (cost=59119.83 rows=670) (actual time=9.430..1778.506 rows=6459 loops=1)
            -> Filter: (RetentionPoliciesChannels.ChannelId is null)  (cost=59119.83 rows=670) (actual time=9.361..1763.193 rows=6459 loops=1)
                -> Nested loop antijoin  (cost=59119.83 rows=670) (actual time=9.356..1762.293 rows=6459 loops=1)
                    -> Inner hash join (c.TeamId = Teams.Id)  (cost=25562.85 rows=670) (actual time=3.940..1742.152 rows=6459 loops=1)
                        -> Filter: (c.`Type` in ('O','P'))  (cost=2489.19 rows=6698) (actual time=2.822..1735.645 rows=6459 loops=1)
                            -> Index range scan on c using idx_channels_delete_at over (DeleteAt = 0), with index condition: (c.DeleteAt = 0)  (cost=2489.19 rows=133960) (actual time=0.975..1687.671 rows=270959 loops=1)
                        -> Hash
                            -> Table scan on Teams  (cost=1.20 rows=2) (actual time=1.051..1.073 rows=2 loops=1)
                    -> Single-row covering index lookup on RetentionPoliciesChannels using PRIMARY (ChannelId=c.Id)  (cost=0.50 rows=1) (actual time=0.003..0.003 rows=0 loops=6459)
 |
```

NEW
```
| -> Limit: 10 row(s)  (actual time=1225.650..1225.656 rows=10 loops=1)
    -> Sort: c.DisplayName, Teams.DisplayName, limit input to 10 row(s) per chunk  (actual time=1225.649..1225.654 rows=10 loops=1)
        -> Stream results  (cost=61276.95 rows=66980) (actual time=1.352..1220.833 rows=6459 loops=1)
            -> Nested loop left join  (cost=61276.95 rows=66980) (actual time=1.338..1209.904 rows=6459 loops=1)
                -> Nested loop antijoin  (cost=46206.45 rows=66980) (actual time=1.331..1200.858 rows=6459 loops=1)
                    -> Nested loop inner join  (cost=32810.45 rows=66980) (actual time=1.273..1195.289 rows=6459 loops=1)
                        -> Table scan on Teams  (cost=0.45 rows=2) (actual time=0.043..0.048 rows=2 loops=1)
                        -> Filter: ((c.TeamId = Teams.Id) and (c.`Type` in ('O','P')))  (cost=4683.50 rows=33490) (actual time=0.708..597.414 rows=3230 loops=2)
                            -> Index lookup on c using idx_channels_delete_at (DeleteAt=0)  (cost=4683.50 rows=133960) (actual time=0.210..566.134 rows=270959 loops=2)
                    -> Single-row index lookup on <subquery2> using <auto_distinct_key> (ChannelId=c.Id)  (actual time=0.001..0.001 rows=0 loops=6459)
                        -> Materialize with deduplication  (cost=0.45..0.45 rows=1) (actual time=0.049..0.049 rows=0 loops=1)
                            -> Filter: (RetentionPoliciesChannels.ChannelId is not null)  (cost=0.35 rows=1) (actual time=0.047..0.047 rows=0 loops=1)
                                -> Covering index scan on RetentionPoliciesChannels using IDX_RetentionPoliciesChannels_PolicyId  (cost=0.35 rows=1) (actual time=0.046..0.046 rows=0 loops=1)
                -> Single-row covering index lookup on RetentionPoliciesChannels using PRIMARY (ChannelId=c.Id)  (cost=0.13 rows=1) (actual time=0.001..0.001 rows=0 loops=6459)
 |
```

I am going to drop this one. Given the query structure, we have to keep the left join.

**Comment by agnivade** [comment]
Closing this as keeping the left join as even worse characteristics.
